### PR TITLE
fix: Metadata UI improvements

### LINKED ISSUE #4295: Errors when saving feature metadata fields are silently ignored
### How are you running Flagsmith

- [ ] Self Hosted with Docker
- [ ] Self Hosted with Kubernetes
- [X] SaaS at flagsmith.com
- [ ] Some other way (add details in description below)

### Describe the bug

When saving any of these changes to metadata in a feature, no error is shown and the changes are not persisted after refreshing the page:

* Erasing a required field and leaving it blank
* Setting an int field to a string
* Setting a URL field to an invalid URL

### Steps To Reproduce

1. Create a required feature metadata field
2. Create a feature and provide the required field
3. Edit the feature, and remove the value for the metadata field
4. Changes are saved, but after you refresh the page the old value is still there

### Expected behavior

Users should not be allowed to save changes to features that have invalid metadata fields (invalid type or missing value on a required field).

### Screenshots

https://github.com/Flagsmith/flagsmith/assets/829698/d88435a4-7141-4dd1-a612-ffa32fd8bc27

### LINKED ISSUE #4303: mutiline string metadata fields behave identically to string fields
### How are you running Flagsmith

- [ ] Self Hosted with Docker
- [ ] Self Hosted with Kubernetes
- [X] SaaS at flagsmith.com
- [ ] Some other way (add details in description below)

### Describe the bug

Creating a metadata field of type "multiline string" does not allow you to type or paste newlines, meaning that it behaves the same way as regular string fields.

### Steps To Reproduce

Edit any multiline string field in features, segments or environments.

### Expected behavior

Display a textarea control when editing multiline string fields and allow users to input or paste newlines.

FWIW, I don't actually know if there's a valid use case for multiline string fields. I would also be okay with hiding them for now.

### Screenshots

_No response_

### LINKED ISSUE #4306: Inconsistent client-side and server-side validation for URL metadata fields
### How are you running Flagsmith

- [ ] Self Hosted with Docker
- [ ] Self Hosted with Kubernetes
- [X] SaaS at flagsmith.com
- [ ] Some other way (add details in description below)

### Describe the bug

URL metadata fields are validated serverside to have a scheme and authority per https://github.com/Flagsmith/flagsmith/blob/1dfcdc2e4c4428110749dac2402e1560c39dedda/api/metadata/models.py#L60

However, client-side a scheme is not mandated, and an error is not shown in this case.

### Steps To Reproduce

1. Save a URL metadata field with a valid value like `https://example.com`
2. Try to update the value to `example.com`

Changes are not saved because the API rejects a URL without a scheme like `example.com` as invalid, but the UI doesn't mark the same value as invalid. Other invalid values like `example` are correctly rejected by both API and UI.

### Expected behavior

Validation for URLs should behave the same way in the UI and API. I would prefer keeping the API's validation logic to always make URL schemes explicit in whatever custom field data we're storing.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by novakzaballa**
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have added information to `docs/` if required so people know about the feature!
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Add error messages and validations in the features and environments settings.
- Improve multiline type metadata.
- Change URL validation type.
- Only enterprise Orgs can use custom fields/metadata

## How did you test this code?

- Use an enterprise organization to be able to use custom fields/metadata.
- Changes were applied to multiline metadata in the feature settings.
- Try to cause an error to see error toasts. For example, enter an incorrect metadata value and try to save.

### CODE REVIEW

File: frontend/common/stores/project-store.js [hunk 1]
```diff
@@ -84,15 +84,22 @@ const controller = {
...
+        store.model.environments[index] = res
+        store.saved()
+        getStore().dispatch(
+          environmentService.util.invalidateTags(['Environment']),
+        )
+        AppActions.refreshOrganisation()
+      })
+      .catch((e) => {
+        e.json().then((result) => {
+          toast(result.metadata[0], 'danger')
```
Comment (kyle-ssg) [inline frontend/common/stores/project-store.js]: This assumes result.metadata isn't undefined.

File: frontend/web/components/metadata/AddMetadataToEntity.tsx [hunk 2]
```diff
@@ -334,7 +335,35 @@ const MetadataRow: FC<MetadataRowType> = ({
...
+              setMetadataValue(!metadataValue)
+              setMetadataValueChanged(true)
+              saveMetadata()
+            }}
+          />
+        </Flex>
+      ) : metadata?.type === 'multiline_str' ? (
+        <Flex className='flex-row' style={{ minWidth: '300px' }}>
+          <InputGroup
+            textarea
```
Comment (kyle-ssg) [inline frontend/web/components/metadata/AddMetadataToEntity.tsx]: The component under this is the exact same except it has a textarea prop, why not just do the following?

```
textarea={metadata?.type === 'multiline_str'}
```

File: frontend/web/components/modals/CreateFlag.js [hunk 3]
```diff
@@ -1961,12 +1961,20 @@ const FeatureProvider = (WrappedComponent) => {
...
         FeatureListStore,
         'saved',
-        ({ changeRequest, createdFlag, isCreate } = {}) => {
-          toast(
-            `${createdFlag || isCreate ? 'Created' : 'Updated'} ${
-              changeRequest ? 'Change Request' : 'Feature'
-            }`,
-          )
+        ({ changeRequest, createdFlag, error, isCreate } = {}) => {
+          if (error?.data?.metadata) {
```
Comment (kyle-ssg) [inline frontend/web/components/modals/CreateFlag.js:1961]: error could be defined but not error?.data?.metadata. With this code any other error will show a success message.

File: frontend/common/stores/project-store.js (hunk 1 above)
Comment (novakzaballa) [reply frontend/common/stores/project-store.js]: Corrected

File: frontend/web/components/metadata/AddMetadataToEntity.tsx (hunk 2 above)
Comment (novakzaballa) [reply frontend/web/components/metadata/AddMetadataToEntity.tsx]: Updated

File: frontend/web/components/modals/CreateFlag.js (hunk 3 above)
Comment (novakzaballa) [reply frontend/web/components/modals/CreateFlag.js:1961]: Updated

File: frontend/common/utils/utils.tsx [hunk 4]
```diff
@@ -382,6 +382,10 @@ const Utils = Object.assign({}, require('./base/_utils'), {
         valid = isEnterprise
         break
       }
+      case 'METADATA': {
```
Comment (kyle-ssg) [inline frontend/common/utils/utils.tsx:385]: Note to self: this affects https://github.com/Flagsmith/flagsmith/pull/4281

File: frontend/common/stores/project-store.js [hunk 5]
```diff
@@ -84,15 +84,30 @@ const controller = {
...
+        store.saved()
+        getStore().dispatch(
+          environmentService.util.invalidateTags(['Environment']),
+        )
+        AppActions.refreshOrganisation()
+      })
+      .catch((e) => {
+        e.json()
+          .then((result) => {
+            if (result.metadata && result.metadata[0]) {
```
Comment (kyle-ssg) [inline frontend/common/stores/project-store.js]: This is better syntax and also checks that result.metadata isn't null (maybe not necessary but still good) 

```
(result: Environment) => { // Types Environment
   if(result?.metadata?.[0])
}
```

File: frontend/common/utils/utils.tsx [hunk 6]
```diff
@@ -396,6 +400,8 @@ const Utils = Object.assign({}, require('./base/_utils'), {
         : null
       : AccountStore.getPlans()
 
+    console.log('DEBUG: plans:', plans)
```
Comment (kyle-ssg) [inline frontend/common/utils/utils.tsx]: can remove

Comment (novakzaballa) [reply frontend/common/utils/utils.tsx]: Done

File: frontend/common/stores/project-store.js (hunk 5 above)
Comment (novakzaballa) [reply frontend/common/stores/project-store.js]: done

**Comment by kyle-ssg** [comment]
As discussed, raised an overall issue [here](https://github.com/Flagsmith/flagsmith/issues/4408). For now there's a few things to tidy up + default custom fields to expanded

**Comment by novakzaballa** [comment]
> As discussed, raised an overall issue [here](https://github.com/Flagsmith/flagsmith/issues/4408). For now there's a few things to tidy up + default custom fields to expanded

@kyle-ssg, I adjusted the Vertical alignment, removed the placeholder, and custom fields expanded by default.
I added a screenshot with the UI changes bellow.

Screenshot 2024-07-29 at 11 34 43 AM

**Review by kyle-ssg — APPROVED** [review]

**Comment by kyle-ssg** [comment]
So much better

**Pull request merged**